### PR TITLE
GEOMESA-1239 Support shpfile ingest over S3 and HDFS

### PR DESCRIPTION
**Pull request by jnh5y**
Signed-off-by: Jim Hughes <jnh5y@ccri.com>

### CODE REVIEW

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/accumulo/commands/IngestCommand.scala [hunk 1]
```diff
@@ -24,12 +25,17 @@ import org.locationtech.geomesa.tools.common.{CLArgResolver, OptionalFeatureType
...
 class IngestCommand(parent: JCommander) extends Command(parent) with LazyLogging {
   override val command = "ingest"
   override val params = new IngestParameters()
 
+  private val remotePrefixes = Seq("hdfs", "s3n", "s3a")
+
+  def isDistributedUrl(url: String) = remotePrefixes.exists(url.startsWith)
+
   override def execute(): Unit = {
     ensureSameFs(Seq("hdfs", "s3n", "s3a"))
```
Comment (jahhulbert-ccri) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/accumulo/commands/IngestCommand.scala]: reuse the val remotePrefixes you pulled out here?

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/accumulo/ingest/AbstractIngest.scala [hunk 2]
```diff
@@ -94,6 +94,18 @@ abstract class AbstractIngest(val dsParams: Map[String, String],
...
   }
 
+  def getTypeAndFeatures(file: File, failed: AtomicLong): (SimpleFeatureType, Iterator[SimpleFeature], CountingInputStream, Closeable) = {
+    val converter = createLocalConverter(file, failed)
+    // count the raw bytes read from the file, as that's what we based our total on
+    val countingStream = new CountingInputStream(new FileInputStream(file))
+    val is = PathUtils.handleCompression(countingStream, file.getPath)
+    val (sft, features) = converter.convert(is)
+    val closeable = new Closeable {
+      override def close(): Unit = { converter.close(); is.close() }
```
Comment (jahhulbert-ccri) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/accumulo/ingest/AbstractIngest.scala]: you need to put this in closeQuietly like before to ensure the is.close() happens...

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/accumulo/ingest/AbstractIngest.scala [hunk 3]
```diff
@@ -107,16 +119,14 @@ abstract class AbstractIngest(val dsParams: Map[String, String],
           // only create the feature writer after the converter runs
           // so that we can create the schema based off the input file
           var fw: FeatureWriter[SimpleFeatureType, SimpleFeature] = null
-          val converter = createLocalConverter(file, failed)
-          // count the raw bytes read from the file, as that's what we based our total on
-          val countingStream = new CountingInputStream(new FileInputStream(file))
-          val is = PathUtils.handleCompression(countingStream, file.getPath)
+
+          val (sft, features, countingStream, closer) = getTypeAndFeatures(file, failed)
```
Comment (jahhulbert-ccri) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/accumulo/ingest/AbstractIngest.scala]: so technically you moved the opening of the input stream outside of the try {} finally {} block need to move it back in

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/GeneralShapefileIngest.scala [hunk 4]
```diff
@@ -24,15 +26,26 @@ object GeneralShapefileIngest {
...
 
+  // The goal of this method is to allow for URL-based look-ups.
+  //  This allows for us to ingest files from HDFS and S3.
+  def getShapefileDatastore(shapefilePath: String): FileDataStore = {
+    if (shapefilePath.contains(":")) {
+      DataStoreFinder.getDataStore(Map("url" -> shapefilePath)).asInstanceOf[FileDataStore]
+    } else {
+      FileDataStoreFinder.getDataStore(new File(shapefilePath))
+    }
+
```
Comment (jahhulbert-ccri) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/GeneralShapefileIngest.scala]: picky - extra newline

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/GeneralShapefileIngest.scala [hunk 5]
```diff
@@ -24,15 +26,26 @@ object GeneralShapefileIngest {
...
     val features = shapefile.getFeatureSource.getFeatures
     val newDS = featuresToDataStoreViaParams(features, params, featureName)
     shapefile.dispose()
     newDS
   }
 
+  // The goal of this method is to allow for URL-based look-ups.
+  //  This allows for us to ingest files from HDFS and S3.
+  def getShapefileDatastore(shapefilePath: String): FileDataStore = {
+    if (shapefilePath.contains(":")) {
```
Comment (jahhulbert-ccri) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/GeneralShapefileIngest.scala]: picky - maybe make a regex looking for more than just ':' match ```<scheme>://<something>```

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/accumulo/commands/IngestCommand.scala [hunk 6]
```diff
@@ -39,7 +45,21 @@ class IngestCommand(parent: JCommander) extends Command(parent) with LazyLogging
...
+
+      // If someone is ingesting file from hdfs or S3, we add the Hadoop URL Factories to the JVM.
+      if (params.files.exists(isDistributedUrl)) {
+        import org.apache.hadoop.fs.FsUrlStreamHandlerFactory
+        val factory = new FsUrlStreamHandlerFactory
+        URL.setURLStreamHandlerFactory(factory)
+      }
+
+      if (params.threads > 1) {
+        val parfiles =  params.files.par
```
Comment (jahhulbert-ccri) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/accumulo/commands/IngestCommand.scala]: extra space

**Comment by jahhulbert-ccri** [comment]
what's the target milestone?

**Comment by jnh5y** [comment]
@jahhulbert-ccri Hmm... it shouldn't hold up 1.2.5, but it'd be a nice-to-have for it.

It obviously isn't ready quite yet though...  Likely to slip.

**Comment by jahhulbert-ccri** [comment]
@jnh5y depending on when we try it shouldn't be too bad. I can add the shape files for testing to my little bash script to test them in hdfs and s3.

**Comment by elahrvivaz** [comment]
looks good to me, pending jahulbert's suggestions

**Comment by jahhulbert-ccri** [comment]
@jnh5y you get a chance to make these changes?

**Comment by jnh5y** [comment]
@jahhulbert-ccri, negative.  @aheyne @cfkelly, maybe one of you gents would be interested and able to take over?

**Comment by aheyne** [comment]
Sure.

**Comment by aheyne** [comment]
Now that I've already pushed a commit up it occurred to me that closeable is being passed back as closer and closeQuietly(closer) is being called so is it necessary to call closeQuietly on converter and is?

**Comment by jahhulbert-ccri** [comment]
so the reason was that those need to be inside the inner try like they were before due to resource handling...the reason it was closeQuietly is bc it is theoretically null

**Comment by jahhulbert-ccri** [comment]
testing and will merge this

**Comment by jahhulbert-ccri** [comment]
fyi i reverted the changes for the 4-tuple method for opening converter, etc bc it doesn't close resources and I don't think it actually was relevant to remote shapefile ingest? someone correct me if i was wrong and we can rework it.

**Pull request merged**